Sources/ShopifyCheckoutSheetKit/PixelEvents.swift

```diff
@@ -36,20 +36,75 @@ public enum PixelEvent {
 	case paymentInfoSubmitted(PixelEventsPaymentInfoSubmitted)
 }
 
+public struct BaseEvent<T>: Codable where T: Codable {
+    public let context: Context?
+    public let data: T?
+    /// The ID of the customer event
+    public let id: String?
+    /// The name of the customer event
+    public let name: String?
+    /// The timestamp of when the customer event occurred, in [ISO
+    /// 8601](https://en.wikipedia.org/wiki/ISO_8601) format
+    public let timestamp: String?
+
+    enum CodingKeys: String, CodingKey {
+        case context, data, id, name, timestamp
+    }
+}
+
+/// The `checkout_address_info_submitted` event logs an instance of a customer
+/// submitting their mailing address. This event is only available in checkouts
+/// where checkout extensibility for customizations is enabled
+public typealias PixelEventsCheckoutAddressInfoSubmitted = BaseEvent<PixelEventsCheckoutAddressInfoSubmittedData>
+
+/// The `checkout_completed` event logs when a visitor completes a purchase. This
+/// event is available on the order status and checkout pages
+public typealias PixelEventsCheckoutCompleted = BaseEvent<PixelEventsCheckoutCompletedData>
+
+/// The `checkout_contact_info_submitted` event logs an instance where a customer
+/// submits a checkout form. This event is only available in checkouts where
+/// checkout extensibility for customizations is enabled
+public typealias PixelEventsCheckoutContactInfoSubmitted = BaseEvent<PixelEventsCheckoutContactInfoSubmittedData>
+
+/// The `checkout_shipping_info_submitted` event logs an instance where the
+/// customer chooses a shipping rate. This event is only available in checkouts
+/// where checkout extensibility for customizations is enabled
+public typealias PixelEventsCheckoutShippingInfoSubmitted = BaseEvent<PixelEventsCheckoutShippingInfoSubmittedData>
+
+/// The `checkout_started` event logs an instance of a customer starting
+/// the checkout process. This event is available on the checkout page. For
+/// checkout extensibility, this event is triggered every time a customer
+/// enters checkout. For non-checkout extensible shops, this event is only
+/// triggered the first time a customer enters checkout.
+public typealias PixelEventsCheckoutStarted = BaseEvent<PixelEventsCheckoutStartedData>
+
+/// The `page_viewed` event logs an instance where a customer visited a page.
+/// This event is available on the online store, checkout, and order status pages
+public typealias PixelEventsPageViewed = BaseEvent<PixelEventsPageViewedData>
+
+/// The `payment_info_submitted` event logs an instance of a customer
+/// submitting their payment information. This event is available on the
+/// checkout page
+public typealias PixelEventsPaymentInfoSubmitted = BaseEvent<PixelEventsPaymentInfoSubmittedData>
+
+/// This event represents any custom events emitted by partners or merchants via
+/// the `publish` method
+public typealias CustomEvent = BaseEvent<CustomData>
```

Comment:
As far as I can tell, the events with "custom data" only seem to contain an additional (optional) `checkout` field. Can we just add this optional field to the BaseEvent and drop the PixelEvent enum since, at that point, all of the events will map to the BaseEvent

Reply:
No, custom events do not contain a `checkout` field. The can contain anything.

Reply:
I think we can replace the `enum` with a `protocol` to make it easier to read the events, but I don't think we can just drop the `enum` an use `BaseEvent` as it is a generic `struct`.

Reply:
That solution would make it more difficult to add more PixelEvents in the future. If we start receiving another event that doesn't have the `checkout` field and has a different one we would have to add it to `PixelEvent`. 

With the current solution we have better typing. For example, `PixelEventsPageViewedData` does not have the `checkout` field and it's clear with the current types. It would be confusing to have the same `data` field for every Web Pixel event.

Reply:
No, it's not really optional, it's required but it was decided by the team to make it optional in the Swift types.

> I don't see why it would be a problem for one type to have it and another to not.

For example, `PixelEventsCheckoutAddressInfoSubmitted` has a `checkout` field but `PixelEventsPageViewed` does not have it.

When you are reading the events with the current solution, if you receive a `PixelEventsPageViewed` event Swift won't let you try to read the `checkout` field. 

And as I said, if we add more events in the future, which can happen, we would end up with a "Standard event type" combining all the fields of every event.

Reply:
N.b. We have `dom` events on their way which have different attributes in the data field (things like the element that was clicked, focused, etc). I believe we'll be proxying them to the sdk when they're available.  Plus the PageViewed exception that Jose Miguel mentioned.

The generics seem to give a bit of flexibility (we could also potentially have a third struct for dom events).

Is there much impact on client code?

Reply:
> When you are reading the events with the current solution, if you receive a PixelEventsPageViewed event Swift won't let you try to read the checkout field.

In the case where you receive a `PixelEventsCheckoutAddressInfoSubmitted` event, you would still have to unwrap the optional to read it, so the same code can be used for both.

```swift
if let checkoutData = event.checkout {
  do something
}
```

![image](https://github.com/Shopify/checkout-sheet-kit-swift/assets/2034704/047d7b9d-4fdf-4518-b7ce-bfd7ec86006a)

Reply:
> We have dom events on their way which have different attributes in the data field

@kiftio do you have an idea of the shape of  `data` on these events?

Reply:
Yeah, that would be nice. I think we've landed here, as the events that were recently dropped had different attributes

So, I suppose the options are:
1. drop generics, have a single data type in the interface with optional checkout attribute:
- nicer interface for clients
- less flexible for new events (with varying attributes)

2. keep generics
- requires an additional switch for standard event types 
- more flexible for any new events (with varying attributes)